consensus/src/experimental/tests/commit_phase_tests.rs

```diff
@@ -0,0 +1,642 @@
+
+fn test_commit_phase_check_commit() {
+    let mut runtime = consensus_runtime();
+    let (
+        _commit_tx,
+        _msg_tx,
+        _commit_result_rx,
+        _self_loop_rx,
+        _safety_rules_container,
+        signers,
+        state_computer,
+        _validator,
+        mut commit_phase,
+    ) = prepare_commit_phase();
+
+    timed_block_on(&mut runtime, async move {
+        let _signer = &signers[0];
+
+        let (vecblocks, li_sig) = prepare_executed_blocks(&signers[0], state_computer);
+
+        assert!(commit_phase.blocks().is_none());
+
+        // when blocks is none
+        commit_phase.check_commit().await.ok();
+
+        assert!(commit_phase.blocks().is_none());
+
+        commit_phase.set_blocks(Some(PendingBlocks::new(vecblocks.clone(), li_sig.clone())));
+
+        // when blocks is good
+        commit_phase.check_commit().await.ok();
+
+        // the block should be consumed
+        assert!(commit_phase.blocks().is_none());
+        assert_eq!(commit_phase.load_back_pressure(), 1);
+
+        // when block contains bad signatures
+        let ledger_info_with_no_sig = LedgerInfoWithSignatures::new(
+            LedgerInfo::new(
+                vecblocks.last().unwrap().block_info(),
+                li_sig.ledger_info().consensus_data_hash(),
+            ),
+            BTreeMap::<AccountAddress, Ed25519Signature>::new(), //empty
+        );
+        commit_phase.set_blocks(Some(PendingBlocks::new(vecblocks, ledger_info_with_no_sig)));
+        commit_phase.check_commit().await.ok();
+
+        // the block should be there
+        assert!(commit_phase.blocks().is_some());
+    });
+}
+
+fn test_commit_phase_process_executed_blocks() {
+    let mut runtime = consensus_runtime();
+    let (
+        _commit_tx,
+        _msg_tx,
+        _commit_result_rx,
+        _self_loop_rx,
+        _safety_rules_container,
+        signers,
+        state_computer,
+        _validator,
+        mut commit_phase,
+    ) = prepare_commit_phase();
+
+    timed_block_on(&mut runtime, async move {
+        let _signer = &signers[0];
+
+        let (vecblocks, li_sig) = prepare_executed_blocks(&signers[0], state_computer);
+
+        let ledger_info_with_no_sig = LedgerInfoWithSignatures::new(
+            LedgerInfo::new(
+                vecblocks.last().unwrap().block_info(),
+                li_sig.ledger_info().consensus_data_hash(),
+            ),
+            BTreeMap::<AccountAddress, Ed25519Signature>::new(), //empty
+        );
+
+        // no signatures
+        assert!(matches!(
+            commit_phase
+                .process_executed_blocks(vecblocks, ledger_info_with_no_sig)
+                .await,
+            Err(_),
+        ));
+    });
+}
+
+fn test_commit_phase_fn() {
+    test_commit_phase_process_commit_vote();
+    test_commit_phase_process_commit_decision();
+    test_commit_phase_check_commit();
+    test_commit_phase_process_executed_blocks();
+}
+
+#[test]
+fn test_commit_phase() {
```

Comment:
it's better to add #[test] to each individual test instead of aggregating them into a single one, safetyrules is anti-pattern because it tries to run those tests with different flags

Reply:
got it. resolved